---
title: "ALMANAC_modelling"
output: html_document
---
```{r setup, include=FALSE}
library(tidyverse)
library(here)
library(tidymodels)
library(tictoc)
library(doParallel)
library(patchwork)
library(ROCR)
library(reticulate)
library(vip)
library(recipeselectors)
library(conflicted)
library(Metrics)
library(ggridges)
library(DarkKinaseTools)

conflict_prefer("slice", "dplyr")
conflict_prefer("filter", "dplyr")
conflict_prefer("rmse", "Metrics")
conflict_prefer("vi", "vip")
knitr::opts_knit$set(root.dir = here())
```

```{r}
#read in data
lasso_selected_features = read_csv(here('results/ALMANAC_klaeger_models/lasso_selected_features_truncated.csv')) %>% 
  mutate(feature_type = if_else(feature_type == "Baseline Gene Expression", "Expression", feature_type))

gene_classifications = read_csv(here('data/klaeger_manual_gene_labels.csv')) %>%
	bind_rows(all_kinases %>% select(symbol,class) %>% rename(gene_name = symbol)) %>% 
	#adding extra label for abbreviated activation name
	bind_rows(data.frame(gene_name = "CSNK2A(1/3)", class = "Light")) %>%
	distinct() %>% 
	mutate(gene_name = str_replace(gene_name, "[-;]", "_")) %>% 
	mutate(gene_name = if_else(
		gene_name == 'HIST2H2BE_HIST1H2BB;HIST1H2BO;HIST1H2BJ;HIST3H2BB;HIST1H2BA',
		'HIST2H2BE_HIST1H2BB_HIST1H2BO_HIST1H2BJ_HIST3H2BB_HIST1H2BA',
		gene_name
	))

exp_interactor_type = read_rds(here('data/expression_feature_interactors.rds')) %>%
	mutate(Variable = paste0(type,"_",hgnc)) %>%
	mutate(interaction = case_when(
		direct_count >= 1 ~ "Inhibition\nInteractor",
		T ~ "Noninteractor"
	)) 
```

```{r}
#preprocessing for figure
lasso_selected_features_medians = lasso_selected_features %>% 
	group_by(feature_type) %>% 
	summarise(median = median(Importance, na.rm = TRUE)) %>% 
	arrange(desc(median))

lasso_selected_features_summary = lasso_selected_features %>% 
	mutate(feature_type = as.factor(feature_type)) %>%  
	mutate(feature_type = 
				 	fct_relevel(feature_type, 
				 							lasso_selected_features_medians$feature_type[1], 
				 							lasso_selected_features_medians$feature_type[2], 
				 							lasso_selected_features_medians$feature_type[3], 
				 							lasso_selected_features_medians$feature_type[4] ))
```

```{r}
lasso_selected_features_summary %>% 
  slice(1:150) %>% 
	ggplot(aes(x = Importance, y = feature_type, fill = factor(stat(quantile)))) +
	 stat_density_ridges(
    geom = "density_ridges_gradient", calc_ecdf = TRUE,
    quantiles = 4, quantile_lines = TRUE
  ) +
  scale_fill_viridis_d(name = "Quartiles") +
	theme_ridges(center_axis_labels = TRUE) + 
	labs(title = " ",x = "LASSO Coefficient value", y = "Data Type") +
	coord_cartesian(clip = "off")
	
ggsave(here('figures/ALMANAC_klaeger/lasso_selected_features_ridgeplot.png'), height = 6, width = 21, units = "cm")
```

```{r}
importance_values_processed = lasso_selected_features_summary %>% 
	group_by(feature) %>% 
	summarise(mean_importance_value = mean(abs(Importance))) %>% 
	arrange((mean_importance_value)) %>%
	mutate(feature_type = case_when(
			str_detect(feature, "^act_") ~ "Inhibition",
			str_detect(feature, "^exp_") ~ "Expression",
			str_detect(feature, "^dep_") ~ "Depmap",
			str_detect(feature, "^cnv_") ~ "CNV",
			str_detect(feature, "^prot_") ~ "Proteomics",
			T ~ feature
		))

importance_inhibition = importance_values_processed %>% 
	filter(feature_type == "Inhibition") %>% 
	mutate(feature = ifelse(feature == "act_CSNK2A1_CSNK2A3","act_CSNK2A(1/3)",feature)) %>%
	extract(feature, c(NA,"gene"),"(.*)_(.*)",remove=F) %>%
	left_join(gene_classifications, by=c('gene'='gene_name')) %>%
	mutate(class = ifelse(is.na(class), "Non-kinase",class)) %>% 
	arrange(desc(mean_importance_value)) 
	
importance_inhibition %>%
  slice(1:25) %>%
	arrange(mean_importance_value) %>% 
	mutate(gene = fct_inorder(gene)) %>% 
	ggplot(aes(x = mean_importance_value, y = gene, fill = class)) +
	geom_col() +
	labs(y='',fill='',x='Inhibition State Importance') +
	theme(
		legend.title = element_blank(),
		legend.position = "bottom",
		legend.direction = "horizontal",
		legend.margin = margin(t = 0, b = 0, l = 0, r = 0, unit = "cm"),
		panel.background = element_rect(fill = "transparent",colour = NA),
    panel.grid.minor = element_blank(), 
    panel.grid.major = element_blank(),
    plot.background = element_rect(fill = "transparent",colour = NA)
      )
ggsave(here('figures/ALMANAC_klaeger/lasso_selected_features_barplot.png'), height = 10.5, width = 10.5, units = "cm")
```

